src/main/kotlin/org/rust/ide/inspections/RsVariableMutableInspection.kt

```diff
@@ -0,0 +1,52 @@
+package org.rust.ide.inspections
+
+import com.intellij.codeInspection.ProblemsHolder
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiRecursiveElementVisitor
+import com.intellij.psi.util.PsiTreeUtil
+import org.rust.ide.annotator.fixes.RemoveMutableFix
+import org.rust.lang.core.psi.RsBlock
+import org.rust.lang.core.psi.RsExpr
+import org.rust.lang.core.psi.RsPatBinding
+import org.rust.lang.core.psi.RsVisitor
+import org.rust.lang.core.psi.impl.mixin.isMut
+import java.util.*
+
+private fun findBlock(expr: PsiElement) = PsiTreeUtil.getNonStrictParentOfType(expr, RsBlock::class.java)
+
+/**
+ * Finds occurrences in the sub scope of expr, so that all will be replaced if replace all is selected.
+ */
+fun findOccurrences(expr: PsiElement): List<RsExpr> {
+    val visitor = object : PsiRecursiveElementVisitor() {
+        val foundOccurrences = ArrayList<RsExpr>()
+
+        override fun visitElement(element: PsiElement) {
+            if (element is RsExpr && expr.text == element.text) {
+                foundOccurrences.add(element)
+            } else {
+                super.visitElement(element)
+            }
+        }
+    }
+
+    val block = findBlock(expr) ?: return emptyList()
+    block.acceptChildren(visitor)
+    return visitor.foundOccurrences
+}
+
+class RsVariableMutableInspection : RsLocalInspectionTool() {
+    override fun getDisplayName() = "No mutable required"
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean) =
+        object : RsVisitor() {
+            override fun visitPatBinding(o: RsPatBinding) {
+                if (!o.isMut) return
+                val occurrences = findOccurrences(o.identifier)
```

Comment:
You need something different than occurrences here, you need `ReferencesSearch`. See how it is used in `org.rust.ide.annotator.RsTraitLineMarkerProvider` .